### ISSUE TITLE
I use the model to train the CULane dataset, but...

### ISSUE THREAD

**Issue by passion3394** [MaybeShewill-CV/lanenet-lane-detection#72]
I use the model to train the CULane dataset, but there are questions:
1 the GPU memory is used 107M, seems that the gpu didn't work. 
2 the training stops at the 3840th epoch, it is very strange, why?

@MaybeShewill-CV need your help, thanks!

**Comment by MaybeShewill-CV** [comment]
@passion3394 First check if you install tensor flow you version. If you have more than one gnu on your computer you need to adjust the code. About the training process strangely stopping issue
 maybe you can supply more details about that.

**Comment by MaybeShewill-CV** [comment]
@passion3394 Tensorflow GPU version, sorry for the spell mistake

**Comment by passion3394** [comment]
@MaybeShewill-CV Yes, I have checked the tensorflow, and the cpu version is working so that the gpu memory is used about 107M. I will install a gpu version tensorflow and try to train again.

**Comment by passion3394** [comment]
@MaybeShewill-CV yes, I have found that I use the tensorflow-cpu version. But why the training stops at the 3840th epoch?

**Comment by cardwing** [comment]
@passion3394 , I have modified the code to train it on CULane and it works well (however, the F1-measure is not satisfying).

**Comment by MaybeShewill-CV** [comment]
@cardwing I have found that the Adam optimizer can converge more quickly than traditional SGD optimizer but the training process can be quite unstable when using Adam and it can not performs as well as the origin training dataset. I will write the enet base net and test the model's performance recently. And I have found the scan model performs quite well on culane dataset and I will try to reimplement the model in tensorflow if I can:)

**Comment by cardwing** [comment]
@MaybeShewill-CV , this is an interesting finding. I have noticed that SCNN also uses SGD as the optimizer (optim.sgd(feval, self.params, self.optimState)). And I am really looking forward to your implementation of SCNN in tensorflow! Thanks a lot in advance.

**Comment by passion3394** [comment]
@MaybeShewill-CV @cardwing Is that possible for me to inference more than 4 lanes if I use the tusimple dataset or CULane dataset?

**Comment by passion3394** [comment]
@MaybeShewill-CV the retraining with tensorflow-gpu is going well, the following is a part of the  training log:

I1023 10:07:25.923465 11817 train_lanenet.py:287] Epoch: 55949 total_loss= 0.077004 binary_seg_loss= 0.045495 instance_seg_loss= 0.150526 accuracy= 0.449155 mean_cost_time= 0.786267s
I1023 10:07:26.943917 11817 train_lanenet.py:287] Epoch: 55950 total_loss= 0.057430 binary_seg_loss= 0.029403 instance_seg_loss= 0.122825 accuracy= 0.641393 mean_cost_time= 0.598062s
I1023 10:07:28.118716 11817 train_lanenet.py:287] Epoch: 55951 total_loss= 0.039001 binary_seg_loss= 0.024051 instance_seg_loss= 0.073886 accuracy= 0.678086 mean_cost_time= 0.813439s
I1023 10:07:29.255681 11817 train_lanenet.py:287] Epoch: 55952 total_loss= 0.025090 binary_seg_loss= 0.017949 instance_seg_loss= 0.041751 accuracy= 0.779286 mean_cost_time= 0.762145s
I1023 10:07:30.450912 11817 train_lanenet.py:287] Epoch: 55953 total_loss= 0.053331 binary_seg_loss= 0.036227 instance_seg_loss= 0.093241 accuracy= 0.607971 mean_cost_time= 0.805587s
I1023 10:07:31.702351 11817 train_lanenet.py:287] Epoch: 55954 total_loss= 0.081827 binary_seg_loss= 0.036297 instance_seg_loss= 0.188065 accuracy= 0.407057 mean_cost_time= 0.872389s
I1023 10:07:32.808821 11817 train_lanenet.py:287] Epoch: 55955 total_loss= 0.102116 binary_seg_loss= 0.050483 instance_seg_loss= 0.222592 accuracy= 0.406831 mean_cost_time= 0.712411s


My question is that the accuracy is very low(0.406831) even the epoch number is bigger that 55900, is that normal?

**Comment by cardwing** [comment]
@passion3394 , the accuracy seems normal at least in my opinion. However, you need to test the trained model in CULane testing set and check if the F1-measure is satisfying (refer to [this paper](https://arxiv.org/pdf/1712.06080.pdf)).

**Comment by cardwing** [comment]
@MaybeShewill-CV , if you want to write ENet, I strongly recommend [this repo](https://github.com/kwotsin/TensorFlow-ENet).

**Comment by MaybeShewill-CV** [comment]
@cardwing Ok thanks a lot:)

**Comment by cardwing** [comment]
@MaybeShewill-CV , I am working on implementing the Tensorflow version of SCNN and have found that SGD not only performs more stable but also converge faster compared with Adam using the same learning rate scheme (see the following figure, first is adam and second is sgd). That's really interesting. However, currently the data augmentation and message passing module have not been added.
![adam](https://user-images.githubusercontent.com/23656119/47405187-fcb15300-d782-11e8-96cd-710871c59218.png)
![sgd](https://user-images.githubusercontent.com/23656119/47405193-0470f780-d783-11e8-889f-682e2021042a.png)

**Comment by MaybeShewill-CV** [comment]
@cardwing Yep I found the loss value can easily reach nan when using Adam optimizer but it does not happen when using the SGD optimizer. I am wondering if there is something wrong with the tensorflow implementation:) Also looking forward to your future work.

**Comment by cardwing** [comment]
@MaybeShewill-CV , thanks for your encouragement. Besides, I may have found the reason why using different optimizer behaves differently. This is due to the fact that [tf.contrib.layers.batch_norm](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/batch_norm) does not use the scale parameter by default. In this condition, batch normalization cannot fully work and stabilize the training process. I am setting scale to True to see if this does have an effect on the training process.

**Comment by MaybeShewill-CV** [comment]
@cardwing I have upgrade my local tensorflow to 1.10 and use the tf.layers.batchnormalization to replace the old api. Maybe that is the reason why the whole training process becomes more stable:)

**Comment by PenghuahuaPeng** [comment]
@MaybeShewill-CV 您在culane数据集上训练准确率是多少？8w epoch  训练集60-70间正常吗

**Comment by zguo008** [comment]
Hi:) @cardwing 
I have downloaded the CUlane dataset and arrange the dataset in the suggested way. But the training cannot start. there's an error: F ./tensorflow/core/util/cuda_launch_config.h:127] Check failed: work_element_count > 0 (0 vs. 0).

thus, can u kindly tell me how u modify the CUlane dataset to fit this network?

Best regards,

Zixun

**Comment by cardwing** [comment]
Maybe there is something wrong with cuda or tensorflow-gpu.

**Comment by zguo008** [comment]
@cardwing hi:) thank you for sharing your retraining information! I have spent the past 2 weeks in training the model with the culane dataset. However the training results is not very satisfying. I merely change the learning rate. At first I changed it to 0.000125 which is 1/4 of the original learning rate but the loss converges VERY VERY slowly. Then I change it to 0.0004 and 0.0003,  but there is a severe overfitting going on.
![screenshot_2019-03-04 tensorboard](https://user-images.githubusercontent.com/44756093/53705614-b8387100-3e60-11e9-9e38-1660226b61c9.png)

May I know what parameter you are using to achieve such high accuracy? Or did u add dropout in your model? I am just a student trying to messing around with it since I dont have much experience on training data. Any help will be appreciated!

Best regards,

ZX

**Comment by cardwing** [comment]
@zguo008, maybe you can have a try on [this repo](https://github.com/cardwing/Codes-for-Lane-Detection).

**Comment by zguo008** [comment]
@cardwing Thank you very much!!

**Comment by longtong-system** [comment]
> @cardwing hi:) thank you for sharing your retraining information! I have spent the past 2 weeks in training the model with the culane dataset. However the training results is not very satisfying. I merely change the learning rate. At first I changed it to 0.000125 which is 1/4 of the original learning rate but the loss converges VERY VERY slowly. Then I change it to 0.0004 and 0.0003, but there is a severe overfitting going on.
> ![screenshot_2019-03-04 tensorboard](https://user-images.githubusercontent.com/44756093/53705614-b8387100-3e60-11e9-9e38-1660226b61c9.png)
> 
> May I know what parameter you are using to achieve such high accuracy? Or did u add dropout in your model? I am just a student trying to messing around with it since I dont have much experience on training data. Any help will be appreciated!
> 
> Best regards,
> 
> ZX
hello,Can you give me a copy of the code you downloaded at that time? Because the current code has been updated, the accuracy of the call part has been deleted, so I want to refer to the code at that time, thank you
